AttributeError: 'KV' object has no attribute 'set'

```
2017-11-24 18:06:41,235 [ERROR] conjure-up/openstack-novalxd - events.py:205 - Error in cleanup code: 'KV' object has no attribute 'set'
Traceback (most recent call last):
  File "/snap/conjure-up/837/lib/python3.6/site-packages/conjureup/events.py", line 186, in shutdown_watcher
    await app.save()
  File "/snap/conjure-up/837/lib/python3.6/site-packages/conjureup/app_config.py", line 198, in save
    self.state.set(self._internal_state_key, self.to_json())
AttributeError: 'KV' object has no attribute 'set'
```

--- a/conjureup/app_config.py
+++ b/conjureup/app_config.py
@@ -175,7 +175,9 @@ class AppConfig:
     def from_json(self, data):
         """ Deserializes application state and updates app_config
         """
-        state = json.loads(data.decode('utf8'))
+        if isinstance(data, bytes):
+            data = data.decode('utf8')
+        state = json.loads(data)
         for k, v in state.items():
             try:
                 getattr(self, k)
@@ -195,8 +197,7 @@ class AppConfig:
             # Check for existing key and clear it
             self.state.pop(self._internal_state_key, None)
         else:
-            self.state.set(self._internal_state_key, self.to_json())
-            self.state.flush()
+            self.state[self._internal_state_key] = self.to_json()
             self.log.info('State saved')
 
     async def restore(self):
@@ -212,7 +213,7 @@ class AppConfig:
             result = self.state.get(self._internal_state_key)
             if result:
                 self.log.info("Found cached state, reloading.")
-                self.from_json(result.decode('utf8'))
+                self.from_json(result)
         except json.JSONDecodeError as e:
             # Dont fail fatally if state information is incorrect. Just log it
             # and move on

--- a/test/test_app_config.py
+++ b/test/test_app_config.py
@@ -75,7 +75,7 @@ class AppConfigTestCase(unittest.TestCase):
         self.app.juju.authenticated = False
 
         with test_loop() as loop:
-            yield self.app.save()
+            loop.run_until_complete(self.app.save())
             loop.run_until_complete(self.app.restore())
 
         results_json = self.app.state.get(self.app._internal_state_key)

--- a/test/test_app_config.py
+++ b/test/test_app_config.py
@@ -79,9 +79,12 @@ class AppConfigTestCase(unittest.TestCase):
             loop.run_until_complete(self.app.restore())
 
         results_json = self.app.state.get(self.app._internal_state_key)
-        results = json.loads(results_json.decode('utf8'))
+        results = json.loads(results_json)
 
-        assert self.app.app.controller == results['controller']
+        # FIXME: provider is currently blacklisted, so we can't save or
+        # restore state for it; adding that will take a bit of work
+        # assert self.app.provider.controller == results['controller']
+        assert self.app.config['spell'] == results['config']['spell']
 
     @unittest.skip("Also need serialize/deserialize during "
                    "save/restore of app.provider class")

